components/encryption/src/manager/mod.rs

```diff
@@ -175,31 +190,80 @@ impl Dicts {
     }
 
     fn get_file(&self, fname: &str) -> FileInfo {
-        let dict = self.file_dict.lock().unwrap();
-        let file_info = dict.files.get(fname);
+        let file_info = self.files_cache.get(fname);
         match file_info {
             None => {
                 // Return Plaintext if file not found
                 let mut file = FileInfo::default();
                 file.method = compat(EncryptionMethod::Plaintext);
                 file
             }
-            Some(info) => info.clone(),
+            Some(info) => info.value().clone(),
         }
     }
 
+    fn check_write_error(has_error: bool) -> Result<()> {
+        if has_error {
+            return Err(Error::Io(IoError::new(
+                ErrorKind::Other,
+                format!("IoError occurs when save file dict"),
+            )));
+        }
+        Ok(())
+    }
+
+    fn pipeline_sync(&self) -> Result<()> {
+        let currenct_id = current().id();
+        let finished = Arc::new(AtomicBool::new(false));
+        let has_err = Arc::new(AtomicBool::new(false));
+        let writers_len = {
+            let mut write_queue = self.write_queue.lock().unwrap();
+            write_queue.push_back((currenct_id, finished.clone(), has_err.clone()));
+            while !write_queue.is_empty() {
+                let leader = write_queue.front().unwrap();
+                if leader.0 == currenct_id {
+                    break;
+                }
+                write_queue = self.log_cond.wait(write_queue).unwrap();
+                if finished.load(Ordering::Acquire) {
+                    return Self::check_write_error(has_err.load(Ordering::Acquire));
+                }
+            }
+            if finished.load(Ordering::Acquire) {
+                return Self::check_write_error(has_err.load(Ordering::Acquire));
+            }
+            assert!(write_queue.len() > 0);
+            write_queue.len()
+        };
+        let ret = self.save_file_dict();
+        let save_err = ret.is_err();
+        if let Err(e) = ret {
+            error!("encryption: failed to save file dictionary."; "error" => ?e);
+        }
+        let mut write_queue = self.write_queue.lock().unwrap();
```

Comment:
The logic is complicated... I read the part several times, it should be okay. @yiwu-arbug Pls check again.

Reply:
I will add some comment to make it easy to understand.